src/main/java/appeng/client/gui/widgets/MEGuiTextField.java

```diff
@@ -31,111 +34,236 @@
  * <p>
  * The rendering does pay attention to the size of the '_' caret.
  */
-public class MEGuiTextField extends GuiTextField {
+public class MEGuiTextField implements ITooltip
+{
+	protected GuiTextField field;
+
 	private static final int PADDING = 2;
+	private static final int MAX_INPUT_LENGTH = 25;
+	private static boolean previousKeyboardRepeatEnabled;
+	private static MEGuiTextField previousKeyboardRepeatEnabledField;
 	private String tooltip;
+	private int fontPad;
 
-	private final int _xPos;
-	private final int _yPos;
-	private final int _width;
-	private final int _height;
-	private final int _fontPad;
+	public int x;
+	public int y;
+	public int w;
+	public int h;
 
 	/**
 	 * Uses the values to instantiate a padded version of a text field.
 	 * Pays attention to the '_' caret.
 	 *
-	 * @param fontRenderer renderer for the strings
-	 * @param xPos         absolute left position
-	 * @param yPos         absolute top position
 	 * @param width        absolute width
 	 * @param height       absolute height
+	 * @param tooltip      tooltip message
 	 */
-	public MEGuiTextField(final FontRenderer fontRenderer, final int xPos, final int yPos, final int width, final int height) {
-		super(fontRenderer, xPos + PADDING, yPos + PADDING, width - 2 * PADDING - fontRenderer.getCharWidth('_'), height - 2 * PADDING);
+	public MEGuiTextField(final int width, final int height, final String tooltip)
```

Comment:
it's a breaking change since you removed the origin constructor. other ae2 addon will crash if it uses MEGuiTextField.

Reply:
which mod is broken?

Reply:
It is not part of teh API so it is not strictly required, but I think there is no harm to preserve compat

Reply:
ec2's fluid terminal gui, better p2p's advance memory card gui. maybe more, need further checking

Reply:
'ec2's fluid terminal gui' and 'advance memory card' do not use this class. The only way to keep backwards compatibility is to create another class. I'm not sure if it needed

Reply:
i check ae2 & extra cells

Reply:
oh sorry i mistake the GuiTextField with the MaeGuiTextField in ec2. but better p2p is actually using that https://github.com/GTNewHorizons/BetterP2P/blob/master/src/main/java/com/projecturanus/betterp2p/client/gui/GuiAdvancedMemoryCard.kt#L76

Reply:
Maybe it's better to change `BetterP2P`? Because changed MEGuiTextField dosn't have `maxStringLength`, `setTextColor`, `setVisible` and ext. Otherwise need to create another class to maintain compatibility

Reply:
well that is fine. i will fix that in better p2p

Reply:
Thanks

Reply:
If you use new `IDropToFillTextField` interface. then D&D from NEI also will be work